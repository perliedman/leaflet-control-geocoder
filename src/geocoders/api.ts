import * as L from 'leaflet';

/**
 * An object that represents a result from a geocoding query
 */
export interface GeocodingResult {
  /**
   * Name of found location
   */
  name: string;
  /**
   * The bounds of the location
   */
  bbox: L.LatLngBounds;
  /**
   * The center coordinate of the location
   */
  center: L.LatLng;
  /**
   * URL for icon representing result; optional
   */
  icon?: string;
  /**
   * HTML formatted representation of the name
   */
  html?: string;
  /**
   * Additional properties returned by the geocoder
   */
  properties?: any;
}

/**
 * A callback function used in {@link IGeocoder.geocode} and {@link IGeocoder.suggest} and {@link IGeocoder.reverse}
 */
export type GeocodingCallback = (result: GeocodingResult[]) => void;

/**
 * An interface implemented to respond to geocoding queries
 */
export interface IGeocoder {
  /**
   * Performs a geocoding query and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  geocode(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a geocoding query suggestion (this happens while typing) and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  suggest?(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a reverse geocoding query and returns the results to the callback in the provided context
   * @param location the coordinate to reverse geocode
   * @param scale the map scale possibly used for reverse geocoding
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  reverse?(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void;
}

export interface GeocoderOptions {
  /**
   * URL of the service
   */
  serviceUrl: string;
  /**
   * Additional URL parameters (strings) that will be added to geocoding requests
   */
  geocodingQueryParams?: Record<string, unknown>;
  /**
   * Additional URL parameters (strings) that will be added to reverse geocoding requests
   */
  reverseQueryParams?: Record<string, unknown>;
  /**
   * API key to use this service
   */
  apiKey?: string;
}

/**
 * @internal
 */
export function geocodingParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.geocodingQueryParams);
}

/**
 * @internal
 */
export function reverseParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.reverseQueryParams);
}
