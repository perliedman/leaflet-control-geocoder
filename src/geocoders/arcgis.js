import L from 'leaflet';
import { getJSON } from '../util';

export default {
  class: L.Class.extend({
    options: {
      service_url: 'http://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer'
    },

    initialize: function(accessToken, options) {
      L.setOptions(this, options);
      this._accessToken = accessToken;
    },

    geocode: function(query, cb, context) {
      var params = {
        SingleLine: query,
        outFields: 'Addr_Type',
        forStorage: false,
        maxLocations: 10,
        f: 'json'
      };

      if (this._key && this._key.length) {
        params.token = this._key;
      }

      getJSON(this.options.service_url + '/findAddressCandidates', params, function(data) {
        var results = [],
          loc,
          latLng,
          latLngBounds;

        if (data.candidates && data.candidates.length) {
          for (var i = 0; i <= data.candidates.length - 1; i++) {
            loc = data.candidates[i];
            latLng = L.latLng(loc.location.y, loc.location.x);
            latLngBounds = L.latLngBounds(
              L.latLng(loc.extent.ymax, loc.extent.xmax),
              L.latLng(loc.extent.ymin, loc.extent.xmin)
            );
            results[i] = {
              name: loc.address,
              bbox: latLngBounds,
              center: latLng
            };
          }
        }

        cb.call(context, results);
      });
    },

    suggest: function(query, cb, context) {
      return this.geocode(query, cb, context);
    },

    reverse: function(location, scale, cb, context) {
      var params = {
        location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
        distance: 100,
        f: 'json'
      };

      getJSON(this.options.service_url + '/reverseGeocode', params, function(data) {
        var result = [],
          loc;

        if (data && !data.error) {
          loc = L.latLng(data.location.y, data.location.x);
          result.push({
            name: data.address.Match_addr,
            center: loc,
            bounds: L.latLngBounds(loc, loc)
          });
        }

        cb.call(context, result);
      });
    }
  }),

  factory: function(accessToken, options) {
    return new L.Control.Geocoder.ArcGis(accessToken, options);
  }
};
