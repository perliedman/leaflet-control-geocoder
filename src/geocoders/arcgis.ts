import * as L from 'leaflet';
import { getJSON } from '../util';
import {
  IGeocoder,
  GeocoderOptions,
  GeocodingCallback,
  geocodingParams,
  GeocodingResult,
  reverseParams
} from './api';

export interface ArcGisOptions extends GeocoderOptions {}

/**
 * Implementation of the [ArcGIS geocoder](https://developers.arcgis.com/features/geocoding/)
 */
export class ArcGis implements IGeocoder {
  options: ArcGisOptions = {
    serviceUrl: 'https://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer',
    apiKey: ''
  };

  constructor(options?: Partial<ArcGisOptions>) {
    L.Util.setOptions(this, options);
  }

  geocode(query: string, cb: GeocodingCallback, context?: any): void {
    const params = geocodingParams(this.options, {
      token: this.options.apiKey,
      SingleLine: query,
      outFields: 'Addr_Type',
      forStorage: false,
      maxLocations: 10,
      f: 'json'
    });

    getJSON(this.options.serviceUrl + '/findAddressCandidates', params, data => {
      const results: GeocodingResult[] = [];
      if (data.candidates && data.candidates.length) {
        for (let i = 0; i <= data.candidates.length - 1; i++) {
          const loc = data.candidates[i];
          const latLng = L.latLng(loc.location.y, loc.location.x);
          const latLngBounds = L.latLngBounds(
            L.latLng(loc.extent.ymax, loc.extent.xmax),
            L.latLng(loc.extent.ymin, loc.extent.xmin)
          );
          results[i] = {
            name: loc.address,
            bbox: latLngBounds,
            center: latLng
          };
        }
      }

      cb.call(context, results);
    });
  }

  suggest(query: string, cb: GeocodingCallback, context?: any): void {
    return this.geocode(query, cb, context);
  }

  reverse(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void {
    const params = reverseParams(this.options, {
      location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
      distance: 100,
      f: 'json'
    });
    getJSON(this.options.serviceUrl + '/reverseGeocode', params, data => {
      const result: GeocodingResult[] = [];
      if (data && !data.error) {
        const center = L.latLng(data.location.y, data.location.x);
        const bbox = L.latLngBounds(center, center);
        result.push({
          name: data.address.Match_addr,
          center: center,
          bbox: bbox
        });
      }

      cb.call(context, result);
    });
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link ArcGis}
 * @param options the options
 */
export function arcgis(options?: Partial<ArcGisOptions>) {
  return new ArcGis(options);
}
