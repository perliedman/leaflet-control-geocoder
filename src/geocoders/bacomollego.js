var L = require('leaflet'),
	Util = require('../util'),
	proj4 = require('proj4');

module.exports = {
	class: L.Class.extend({
		options: {
			tmerc: new proj4.Proj("+proj=tmerc +lat_0=-34.629269 +lon_0=-58.463300 +k=0.999998 +x_0=100000 +y_0=100085 +ellps=intl +units=m +no_defs"),
			wsg: new proj4.Proj("EPSG:4326"),
			geocodeUrl: 'https://ws.usig.buenosaires.gob.ar/geocoder/2.2/geocoding/',
			reverseUrl: 'https://ws.usig.buenosaires.gob.ar/geocoder/2.2/reversegeocoding/',
			normalizarUrl: 'https://servicios.usig.buenosaires.gob.ar/normalizar',
			timeout: 10000,
			maxOptions: 3
		},

		initialize: function(options) {
			L.setOptions(this, options);
		},

		_normalizar: function(query, cb, context){
			Util.jsonp(this.options.normalizarUrl, {
				maxOptions: this.options.maxOptions,
				direccion: query
			}, function(json){
				cb.call(context, json);
			}, this, "callback");
			return this;
		},

		_tmerc2LatLng: function(a){
			var b = proj4.toPoint(a);
			return proj4.transform(this.options.tmerc, this.options.wsg, b),
			[ b.y, b.x ];
		},

		geocode: function(query, cb, context) {
			var results = [],
				maxOptions = 0,
				timedOut = false,
				timer;

			function checkIfCb(){
				if(!timedOut && results.length >= maxOptions){
					timedOut = true;
					cb.call(context || cb, results);
					clearTimeout(timer);
				}
			}

			function pushAndCb(c){
				results.push(c);
				checkIfCb();
			}

			this._normalizar(query, function(json){
				timer = setTimeout(L.bind(function(){
					timedOut = true;
					cb.call(context || cb, results);
				}, this), this.options.timeout);

				maxOptions = json.direccionesNormalizadas.length;

				json.direccionesNormalizadas.forEach(function(e){
					if(e.coordenadas){
						var center = L.latLng(e.coordenadas.y, e.coordenadas.x);
						pushAndCb({
							bbox: L.latLngBounds(center, center),
							name: e.direccion,
							center: center
						});
					}else if([
							(e.nombre_localidad||"").toUpperCase(),
							(e.nombre_partido||"").toUpperCase(),
					].indexOf("CABA") > -1)
						L.bind(function(e){
							Util.jsonp(this.options.geocodeUrl, {
								cod_calle: e.cod_calle,
								altura: e.altura
							}, function(tmerc){
								if(typeof(tmerc.x) != "undefined" && typeof(tmerc.y) != "undefined"){
									var center = L.latLng.apply(L, this._tmerc2LatLng([tmerc.x, tmerc.y]));
									pushAndCb({
										bbox: L.latLngBounds(center, center),
										name: e.direccion,
										center: center
									});
								}else{
									maxOptions--;
									checkIfCb();
								}
							}, this, "callback");
						}, this)(e);
					else
						this._normalizar(e.direccion, function(json){
							var data = json.direccionesNormalizadas[0];
							if(data && data.coordenadas){
								var center = L.latLng(data.coordenadas.y, data.coordenadas.x);
								pushAndCb({
									bbox: L.latLngBounds(center, center),
									name: data.direccion,
									center: center
								});
							} else{
								maxOptions--;
								checkIfCb();
							}
						}, this);
				}, this);
			}, this);
		},

		suggest: function(query, cb, context) {
			return this.geocode(query, cb, context);
		},

		reverse: function(location, scale, cb, context) {
			Util.jsonp(this.options.reverseUrl, {
				x: location.lng,
				y: location.lat,
			}, function(json){
				var results = [{
					bbox: L.latLngBounds(location, location),
					name: json.puerta,
					center: location
				}];
				cb.call(context || cb, results);
			}, "callback");
		}
	}),

	factory: function(options) {
		return new L.Control.Geocoder.BAComoLlego(options);
	}
};

