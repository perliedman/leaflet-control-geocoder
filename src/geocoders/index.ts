export * from './api';
export * from './arcgis';
export * from './bing';
export * from './google';
export * from './here';
export * from './latlng';
export * from './mapbox';
export * from './mapquest';
export * from './neutrino';
export * from './nominatim';
export * from './open-location-code';
export * from './opencage';
export * from './pelias';
export * from './photon';
export * from './what3words';
