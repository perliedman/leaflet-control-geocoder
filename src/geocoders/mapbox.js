import L from 'leaflet';
import { getJSON } from '../util';

export default {
  class: L.Class.extend({
    options: {
      serviceUrl: 'https://api.tiles.mapbox.com/v4/geocode/mapbox.places-v1/',
      geocodingQueryParams: {},
      reverseQueryParams: {}
    },

    initialize: function(accessToken, options) {
      L.setOptions(this, options);
      this.options.geocodingQueryParams.access_token = accessToken;
      this.options.reverseQueryParams.access_token = accessToken;
    },

    geocode: function(query, cb, context) {
      var params = this.options.geocodingQueryParams;
      if (
        typeof params.proximity !== 'undefined' &&
        params.proximity.hasOwnProperty('lat') &&
        params.proximity.hasOwnProperty('lng')
      ) {
        params.proximity = params.proximity.lng + ',' + params.proximity.lat;
      }
      getJSON(this.options.serviceUrl + encodeURIComponent(query) + '.json', params, function(
        data
      ) {
        var results = [],
          loc,
          latLng,
          latLngBounds;
        if (data.features && data.features.length) {
          for (var i = 0; i <= data.features.length - 1; i++) {
            loc = data.features[i];
            latLng = L.latLng(loc.center.reverse());
            if (loc.hasOwnProperty('bbox')) {
              latLngBounds = L.latLngBounds(
                L.latLng(loc.bbox.slice(0, 2).reverse()),
                L.latLng(loc.bbox.slice(2, 4).reverse())
              );
            } else {
              latLngBounds = L.latLngBounds(latLng, latLng);
            }
            results[i] = {
              name: loc.place_name,
              bbox: latLngBounds,
              center: latLng
            };
          }
        }

        cb.call(context, results);
      });
    },

    suggest: function(query, cb, context) {
      return this.geocode(query, cb, context);
    },

    reverse: function(location, scale, cb, context) {
      getJSON(
        this.options.serviceUrl +
          encodeURIComponent(location.lng) +
          ',' +
          encodeURIComponent(location.lat) +
          '.json',
        this.options.reverseQueryParams,
        function(data) {
          var results = [],
            loc,
            latLng,
            latLngBounds;
          if (data.features && data.features.length) {
            for (var i = 0; i <= data.features.length - 1; i++) {
              loc = data.features[i];
              latLng = L.latLng(loc.center.reverse());
              if (loc.hasOwnProperty('bbox')) {
                latLngBounds = L.latLngBounds(
                  L.latLng(loc.bbox.slice(0, 2).reverse()),
                  L.latLng(loc.bbox.slice(2, 4).reverse())
                );
              } else {
                latLngBounds = L.latLngBounds(latLng, latLng);
              }
              results[i] = {
                name: loc.place_name,
                bbox: latLngBounds,
                center: latLng
              };
            }
          }

          cb.call(context, results);
        }
      );
    }
  }),

  factory: function(accessToken, options) {
    return new L.Control.Geocoder.Mapbox(accessToken, options);
  }
};
