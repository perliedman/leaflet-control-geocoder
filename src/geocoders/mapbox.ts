import * as L from 'leaflet';
import { getJSON } from '../util';
import {
  IGeocoder,
  GeocoderOptions,
  GeocodingCallback,
  geocodingParams,
  GeocodingResult,
  reverseParams
} from './api';

export interface MapboxOptions extends GeocoderOptions {}

/**
 * Implementation of the [Mapbox Geocoding](https://www.mapbox.com/api-documentation/#geocoding)
 */
export class Mapbox implements IGeocoder {
  options: MapboxOptions = {
    serviceUrl: 'https://api.mapbox.com/geocoding/v5/mapbox.places/'
  };

  constructor(options?: Partial<MapboxOptions>) {
    L.Util.setOptions(this, options);
  }

  _getProperties(loc) {
    const properties = {
      text: loc.text,
      address: loc.address
    };

    for (let j = 0; j < (loc.context || []).length; j++) {
      const id = loc.context[j].id.split('.')[0];
      properties[id] = loc.context[j].text;

      // Get country code when available
      if (loc.context[j].short_code) {
        properties['countryShortCode'] = loc.context[j].short_code;
      }
    }
    return properties;
  }

  geocode(query: string, cb: GeocodingCallback, context?: any): void {
    const params: any = geocodingParams(this.options, {
      access_token: this.options.apiKey
    });
    if (
      params.proximity !== undefined &&
      params.proximity.lat !== undefined &&
      params.proximity.lng !== undefined
    ) {
      params.proximity = params.proximity.lng + ',' + params.proximity.lat;
    }
    getJSON(this.options.serviceUrl + encodeURIComponent(query) + '.json', params, data => {
      const results: GeocodingResult[] = [];
      if (data.features && data.features.length) {
        for (let i = 0; i <= data.features.length - 1; i++) {
          const loc = data.features[i];
          const center = L.latLng(loc.center.reverse());
          let bbox: L.LatLngBounds;
          if (loc.bbox) {
            bbox = L.latLngBounds(
              L.latLng(loc.bbox.slice(0, 2).reverse()),
              L.latLng(loc.bbox.slice(2, 4).reverse())
            );
          } else {
            bbox = L.latLngBounds(center, center);
          }

          results[i] = {
            name: loc.place_name,
            bbox: bbox,
            center: center,
            properties: this._getProperties(loc)
          };
        }
      }

      cb.call(context, results);
    });
  }

  suggest(query: string, cb: GeocodingCallback, context?: any): void {
    return this.geocode(query, cb, context);
  }

  reverse(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void {
    const param = reverseParams(this.options, {
      access_token: this.options.apiKey
    });
    getJSON(
      this.options.serviceUrl +
        encodeURIComponent(location.lng) +
        ',' +
        encodeURIComponent(location.lat) +
        '.json',
      param,
      data => {
        const results: GeocodingResult[] = [];
        if (data.features && data.features.length) {
          for (let i = 0; i <= data.features.length - 1; i++) {
            const loc = data.features[i];
            const center = L.latLng(loc.center.reverse());
            let bbox: L.LatLngBounds;
            if (loc.bbox) {
              bbox = L.latLngBounds(
                L.latLng(loc.bbox.slice(0, 2).reverse()),
                L.latLng(loc.bbox.slice(2, 4).reverse())
              );
            } else {
              bbox = L.latLngBounds(center, center);
            }
            results[i] = {
              name: loc.place_name,
              bbox: bbox,
              center: center,
              properties: this._getProperties(loc)
            };
          }
        }

        cb.call(context, results);
      }
    );
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link Mapbox}
 * @param options the options
 */
export function mapbox(options?: Partial<MapboxOptions>) {
  return new Mapbox(options);
}
