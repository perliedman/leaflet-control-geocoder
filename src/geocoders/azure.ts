import * as L from 'leaflet';
import { getJSON } from '../util';
import { GeocodingResult, IGeocoder } from './api';

export interface AzureMapsOptions {
  apiKey: string; // Azure Maps API Key
  serviceUrl: string; // Optional: Base URL for the Azure Maps API
}

/**
 * Implementation of [Azure Maps Geocoding](https://www.microsoft.com/en-us/maps/azure/location-services/geocoding)
 *
 * https://learn.microsoft.com/en-us/rest/api/maps/search?view=rest-maps-1.0
 */
export class AzureMaps implements IGeocoder {
  private options: AzureMapsOptions = {
    apiKey: '',
    serviceUrl: 'https://atlas.microsoft.com/search'
  };

  constructor(options: Partial<AzureMapsOptions>) {
    L.Util.setOptions(this, options);
    if (!this.options.apiKey) {
      throw new Error('Azure Maps Geocoder requires an API key.');
    }
  }

  /**
   * {@inheritdoc}
   * https://learn.microsoft.com/en-us/rest/api/maps/search/get-search-address?view=rest-maps-1.0&tabs=HTTP
   */
  async geocode(query: string): Promise<GeocodingResult[]> {
    const params = {
      'api-version': '1.0',
      query,
      'subscription-key': this.options.apiKey
    };
    const url = this.options.serviceUrl + '/address/json';
    const data = await getJSON<AzureMapsResponse>(url, params);

    return (data.results || []).map(
      (result): GeocodingResult => ({
        name: result.address.freeformAddress,
        bbox: new L.LatLngBounds(
          [result.viewport.topLeftPoint.lat, result.viewport.topLeftPoint.lon],
          [result.viewport.btmRightPoint.lat, result.viewport.btmRightPoint.lon]
        ),
        center: new L.LatLng(result.position.lat, result.position.lon)
      })
    );
  }

  /**
   * {@inheritdoc}
   * https://learn.microsoft.com/en-us/rest/api/maps/search/get-search-address-reverse?view=rest-maps-1.0&tabs=HTTP
   */
  async reverse(location: L.LatLngLiteral, scale: number): Promise<GeocodingResult[]> {
    const params = {
      'api-version': '1.0',
      query: location.lat + ',' + location.lng,
      'subscription-key': this.options.apiKey
    };
    const url = this.options.serviceUrl + '/address/reverse/json';
    const data = await getJSON<any>(url, params);

    return (data.addresses || []).map(
      (address): GeocodingResult => ({
        name: address.address.freeformAddress,
        bbox: new L.LatLngBounds(
          [address.viewport.topLeftPoint.lat, address.viewport.topLeftPoint.lon],
          [address.viewport.btmRightPoint.lat, address.viewport.btmRightPoint.lon]
        ),
        center: new L.LatLng(location.lat, location.lng)
      })
    );
  }
}

/**
 * [Class factory method](https://leafletjs.com/reference.html#class-class-factories) for {@link Azure}
 * @param options the options
 */
export function azure(options: AzureMapsOptions) {
  return new AzureMaps(options);
}

/**
 * @internal
 */
export interface AzureMapsResponse {
  summary: Summary;
  results: Result[];
}

interface Result {
  type: string;
  id: string;
  score: number;
  address: Address;
  position: Position;
  viewport: Viewport;
  entryPoints: EntryPoint[];
}

interface Address {
  streetNumber: string;
  streetName: string;
  municipalitySubdivision: string;
  municipality: string;
  countrySecondarySubdivision: string;
  countryTertiarySubdivision: string;
  countrySubdivisionCode: string;
  postalCode: string;
  extendedPostalCode: string;
  countryCode: string;
  country: string;
  countryCodeISO3: string;
  freeformAddress: string;
  countrySubdivisionName: string;
}

interface EntryPoint {
  type: string;
  position: Position;
}

interface Position {
  lat: number;
  lon: number;
}

interface Viewport {
  topLeftPoint: Position;
  btmRightPoint: Position;
}

interface Summary {
  query: string;
  queryType: string;
  queryTime: number;
  numResults: number;
  offset: number;
  totalResults: number;
  fuzzyLevel: number;
}
