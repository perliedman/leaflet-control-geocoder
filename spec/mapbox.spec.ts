import { testXMLHttpRequest } from './mockXMLHttpRequest';
import { Mapbox } from '../src/geocoders/mapbox';

describe('L.Control.Geocoder.Mapbox', () => {
  it('geocodes Milwaukee Ave', () => {
    const geocoder = new Mapbox({ apiKey: '0123' });
    const callback = jest.fn();
    testXMLHttpRequest(
      'https://api.mapbox.com/geocoding/v5/mapbox.places/Milwaukee%20Ave.json?access_token=0123',
      {
        type: 'FeatureCollection',
        query: ['825', 's', 'milwaukee', 'ave', 'deerfield', 'il', '60015'],
        features: [
          {
            id: 'address.4356035406756260',
            type: 'Feature',
            place_type: ['address'],
            relevance: 1,
            properties: {},
            text: 'Milwaukee Ave',
            place_name: '825 Milwaukee Ave, Deerfield, Illinois 60015, United States',
            matching_text: 'South Milwaukee Avenue',
            matching_place_name:
              '825 South Milwaukee Avenue, Deerfield, Illinois 60015, United States',
            center: [-87.921434, 42.166602],
            geometry: {
              type: 'Point',
              coordinates: [-87.921434, 42.166602],
              interpolated: true,
              omitted: true
            },
            address: '825',
            context: [
              {
                id: 'neighborhood.287187',
                text: 'Lake Cook Road'
              },
              {
                id: 'postcode.13903677306297990',
                text: '60015'
              },
              {
                id: 'place.5958304312090910',
                wikidata: 'Q287895',
                text: 'Deerfield'
              },
              {
                id: 'region.3290978600358810',
                short_code: 'US-IL',
                wikidata: 'Q1204',
                text: 'Illinois'
              },
              {
                id: 'country.9053006287256050',
                short_code: 'us',
                wikidata: 'Q30',
                text: 'United States'
              }
            ]
          }
        ],
        attribution:
          'NOTICE: Â© 2018 Mapbox and its suppliers. All rights reserved. Use of this data is subject to the Mapbox Terms of Service (https://www.mapbox.com/about/maps/). This response and the information it contains may not be retained. POI(s) provided by Foursquare.'
      },
      () => geocoder.geocode('Milwaukee Ave', callback)
    );

    const feature = callback.mock.calls[0][0][0];
    expect(feature.name).toBe('825 Milwaukee Ave, Deerfield, Illinois 60015, United States');
    expect(feature.center).toStrictEqual({ lat: 42.166602, lng: -87.921434 });
    expect(feature.bbox).toStrictEqual({
      _northEast: { lat: 42.166602, lng: -87.921434 },
      _southWest: { lat: 42.166602, lng: -87.921434 }
    });
    expect(callback.mock.calls).toMatchSnapshot();
  });
});
