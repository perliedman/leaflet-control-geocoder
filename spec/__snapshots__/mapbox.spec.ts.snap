// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`L.Control.Geocoder.Mapbox geocodes Milwaukee Ave 1`] = `
Array [
  Array [
    Array [
      Object {
        "bbox": Object {
          "_northEast": Object {
            "lat": 42.166602,
            "lng": -87.921434,
          },
          "_southWest": Object {
            "lat": 42.166602,
            "lng": -87.921434,
          },
        },
        "center": Object {
          "lat": 42.166602,
          "lng": -87.921434,
        },
        "name": "825 Milwaukee Ave, Deerfield, Illinois 60015, United States",
        "properties": Object {
          "address": "825",
          "country": "United States",
          "countryShortCode": "us",
          "neighborhood": "Lake Cook Road",
          "place": "Deerfield",
          "postcode": "60015",
          "region": "Illinois",
          "text": "Milwaukee Ave",
        },
      },
    ],
  ],
]
`;
